### ISSUE TITLE
A sensible default setting for complex-valued optimisation

### ISSUE THREAD

**Issue by remifan** [google/jax#7357]
## motivation
I have seen an emerging interest in optimizing complex variables: [TF](https://github.com/tensorflow/tensorflow/issues/38541), [Pytorch](https://github.com/pytorch/pytorch/issues/59998), [Optax](https://github.com/deepmind/optax/issues/144). 

In Jax, for some reasons[#4891](https://github.com/google/jax/issues/4891), the convention of the gradient on complex number given map `f: ℂ→ℝ` is in fact the conjugation of steepest descent direction. It imposes some inconvenience in using basic `jax.experimental.optimizer.sgd` out of the box unless rewriting its enclosure. In addition, a widely seen bother is that one needs to adapt a host of optimizers having second-order momentum(e.g., Adam) since the square of grad does not stand for its variance which might translate into unexpected behaviors as demonstrated in the above referred discussions.

## proposal
Replacing `grad * grad` with `grad * grad.conj()` is one of the agreed workarounds yet I can hardly find rigorous studies in the literature. Despite the `grad * grad.conj()` proposal seems working for sensible reasons, there are more real-value restricted operations(e.g., value comparison in _amsgrad_ mode), which renders more incremental researches and surgical modifications. Given those concerns, it is handy to have a universal treatment that makes existing optimizers working as expected on complex variables out of the box from the caller's perspective while leaving no touch on the implementation of existing optimizers for conventional purposes. 

As justified by @Lezcano in [discussion](https://github.com/pytorch/pytorch/issues/59998#issuecomment-880686890),  one can treat a complex variable as 2 real variables, which solely doubles the number of parameters from the perspective of the optimizers. Though it may not be optimal since many complex objects's real and imag parts are connected by some rules. Without knowing any domain-specific implications, the proposed extension should be a minimal superset of the current default. 

## example implementation
Thanks to Jax's nice PyTree utility, the implement is straightforward, here is [my example](https://github.com/remifan/jax/commit/a9ab514b442dff498d82d39a64c49c7dd1457c36#diff-b8da889f8265643ac4f145d572dd7c0f761eab33102aa9d38045e13c98dff3ec)

**Comment by mattjj** [comment]
Thanks for the suggestion, and the clear explanation!

This line of thinking makes sense to me. I like how you factored out `complex_handler`. And I'm glad to hear that the extension turned out to be straightforward!

However, we're not really developing `jax.experimental.optimizers` more. From the [`jax.experimental` README](https://github.com/google/jax/tree/main/jax/experimental#mini-libraries):

> To serve that purpose, it is best to keep their code samples minimal; so we generally will not merge PRs adding new features. Instead, please send your lovely pull requests and design ideas to more fully-featured libraries like Haiku, Flax, or Trax.

Because these changes are somewhat exploratory, I'm inclined to keep `jax.experimental.optimizers` a bit simpler and more canonical by restricting to the real-valued setting. Even though that means one can't use `jax.experimental.optimizers.sgd` out-of-the-box, as you point out, it still wouldn't be hard for users to write a new optimizer specialized to complex-valued decision variables. After all, that's what the code in `jax.experimental.optimizers` is going for: bare-bones, but easy to fork and tweak to your purposes.

So for those reasons I'm inclined not to make this enhancement to `jax.experimental.optimizers`.

What do you think?

**Comment by remifan** [comment]
@mattjj Thanks for clarifying your purpose! Actually, I really like how `jax.experimental.optimizers` is composed in such an essential way that restrains from encompassing incremental enhancements, indeed, I'm benefiting much from this design to shape my own optimizer as you expect!

I suppose the proposed feature may serve as a bare-bones example to extend to complex domain with minimal efforts and assumptions on the basis of the impressive `pytree` utility(without which it would be not that straightforward afaik), which is exactly what `jax.experimental.optimizers` relies on to factor out the optimizer neatly and elegantly.
I'm also aware that the complex handler might be slightly confusing to regular users. But I have seen more ppl from other science disciplines getting into Jax's world, and I see the topic of complex diff has been well [documented](https://jax.readthedocs.io/en/latest/notebooks/autodiff_cookbook.html#complex-numbers-and-differentiation), isn't this extension the last minor step to cater to users dealing with complex values?

Since Jax is more general-purpose compared with its derivatives which haven't covered this topic well yet(they might tend to think this feature is too fundamental), one would likely trace back to Jax for some hints on how to optimize f: ℂ→ℝ instead of doing conjugation and other tweaks. Thus I think there is a meaningful tradeoff to justify this extension in Jax.

I'm looking forward to your suggestion!
